bin/varnishd/cache/cache_vpi.c

```diff
@@ -94,6 +94,11 @@ VPI_vcl_select(VRT_CTX, VCL_VCL vcl)
 	if (IS_TOPREQ(req) && req->top->vcl0 != NULL)
 		return;		// Illegal, req-FSM will fail this later.
 
+	/* XXX VCL_Task* are somewhat duplicated to those in Req_Rollback called
+	 * from FSM for VCL_RET_VCL. Keeping them here to ensure there are no
+	 * tasks during calls to VCL_Rel / vcl_get
+	 */
```

Comment:
I'm not sure I'm comfortable with the need to use `VCL_Task*` twice for the top request. I haven't looked in details yet but maybe we need a `Req_(Enter|Leave)` functions that call `VCL_Task*` twice for each set of PRIVs?

This feels brittle and something we might forgot during a refactoring. Maybe we really need to make the `VCL_Task*` functions static and force ourselves to go through the right `Bo_` or `Req_` variants that would act as wrappers?

Reply:
Yes, we should clear this up. I will issue another PR.